File: hooks/scripts/notifications/lib/env-loader.ts
// Environment loader with cascade: process.env > ~/.claude/.env > .claude/.env
import * as fs from "node:fs";
import * as path from "node:path";
import * as os from "node:os";

/** Parse a .env file content into key-value pairs (supports comments, quoted values) */
export function parseEnvContent(content: string): Record<string, string> {
  const result: Record<string, string> = {};
  for (const line of content.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;

    const eqIndex = trimmed.indexOf("=");
    if (eqIndex === -1) continue;

    const key = trimmed.slice(0, eqIndex).trim();
    let value = trimmed.slice(eqIndex + 1).trim();

    if (
      (value.startsWith('"') && value.endsWith('"')) ||
      (value.startsWith("'") && value.endsWith("'"))
    ) {
      value = value.slice(1, -1);
    } else {
      // Strip inline comments only for unquoted values
      const commentIndex = value.indexOf("#");
      if (commentIndex !== -1) value = value.slice(0, commentIndex).trim();
    }

    if (key) result[key] = value;
  }
  return result;
}

function loadEnvFile(filePath: string): Record<string, string> {
  try {
    if (fs.existsSync(filePath)) {
      const parsed = parseEnvContent(fs.readFileSync(filePath, "utf8"));
      if (Object.keys(parsed).length > 0) {
        process.stderr.write(`[env-loader] Loaded: ${filePath}\n`);
      }
      return parsed;
    }
  } catch (err) {
    process.stderr.write(`[env-loader] Failed to read ${filePath}: ${(err as Error).message}\n`);
  }
  return {};
}

/** Load environment with cascade: .claude/.env (low) → ~/.claude/.env → process.env (high) */
export function loadEnv(cwd = process.cwd()): Record<string, string> {
  const envFiles = [
    path.join(cwd, ".claude", ".env"),
    path.join(os.homedir(), ".claude", ".env"),
  ];

  let merged: Record<string, string> = {};
  for (const filePath of envFiles) {
    merged = { ...merged, ...loadEnvFile(filePath) };
  }
  return { ...merged, ...process.env } as Record<string, string>;
}

File: hooks/scripts/internal/notify/env_loader.go
// Package notify handles notification dispatch to external services.
// env_loader.go: .env file parser with cascade loading.
package notify

import (
	"os"
	"path/filepath"
	"strings"
)

// ParseEnvContent parses .env file content into a key-value map.
// Supports comments (#), quoted values (single/double), inline comments for unquoted values.
func ParseEnvContent(content string) map[string]string {
	result := make(map[string]string)
	for _, line := range strings.Split(content, "\n") {
		trimmed := strings.TrimSpace(line)
		if trimmed == "" || strings.HasPrefix(trimmed, "#") {
			continue
		}

		eqIdx := strings.Index(trimmed, "=")
		if eqIdx == -1 {
			continue
		}

		key := strings.TrimSpace(trimmed[:eqIdx])
		value := strings.TrimSpace(trimmed[eqIdx+1:])

		// Strip surrounding quotes
		if len(value) >= 2 &&
			((value[0] == '"' && value[len(value)-1] == '"') ||
				(value[0] == '\'' && value[len(value)-1] == '\'')) {
			value = value[1 : len(value)-1]
		} else {
			// Strip inline comments for unquoted values
			if cidx := strings.Index(value, "#"); cidx != -1 {
				value = strings.TrimSpace(value[:cidx])
			}
		}

		if key != "" {
			result[key] = value
		}
	}
	return result
}

// loadEnvFile reads and parses a .env file; returns empty map on error.
func loadEnvFile(path string) map[string]string {
	data, err := os.ReadFile(path)
	if err != nil {
		return make(map[string]string)
	}
	parsed := ParseEnvContent(string(data))
	if len(parsed) > 0 {
		os.Stderr.WriteString("[env-loader] Loaded: " + path + "\n")
	}
	return parsed
}

// LoadEnv loads environment with cascade: .claude/.env (low) → ~/.claude/.env → os.Environ (high).
// cwd is used to find the project-local .claude/.env file.
func LoadEnv(cwd string) map[string]string {
	if cwd == "" {
		var err error
		cwd, err = os.Getwd()
		if err != nil {
			cwd = "."
		}
	}

	home, _ := os.UserHomeDir()
	envFiles := []string{
		filepath.Join(cwd, ".claude", ".env"),
		filepath.Join(home, ".claude", ".env"),
	}

	merged := make(map[string]string)
	for _, f := range envFiles {
		for k, v := range loadEnvFile(f) {
			merged[k] = v
		}
	}

	// os.Environ overrides file values
	for _, kv := range os.Environ() {
		idx := strings.Index(kv, "=")
		if idx == -1 {
			continue
		}
		merged[kv[:idx]] = kv[idx+1:]
	}

	return merged
}
